### PR TITLE
[dv/alert_handler] Add timing check for pings

### PR DESCRIPTION
**Pull request by cindychip**
This PR adds more timing check to make sure ping is triggered within the
expected timeframe.

Signed-off-by: Cindy Chen <chencindy@opentitan.org>

### CODE REVIEW

File: hw/ip_templates/alert_handler/dv/env/alert_handler_scoreboard.sv [hunk 1]
```diff
@@ -563,6 +571,38 @@ class alert_handler_scoreboard extends cip_base_scoreboard #(
...
+        fork
+          begin
+            wait (cfg.under_reset == 1);
+            ping_timer_en = 0;
+            num_checked_pings = 0;
+          end
+          begin
+            // The ping timer is 16 bits so ideally we should see alert_ping -> esc_ping ->
+            // alert_ping ... with the max length of 16'hFFFF clock cycle. However alert_ping is
+            // randomly selected and we can not guarantee the random alert index is valid (exists)
```
Comment (sriyerg) [inline hw/ip_templates/alert_handler/dv/env/alert_handler_scoreboard.sv:588]:
```suggestion
            // randomly selected and we can not guarantee the random alert index is valid (exists),
```

File: hw/ip_templates/alert_handler/dv/env/alert_handler_env_cov.sv [hunk 2]
```diff
@@ -281,11 +281,19 @@ class alert_handler_env_cov extends cip_base_env_cov #(.CFG_T(alert_handler_env_
   // Covergroup to make sure simulation is long enough to fetch more than five EDN requests.
   covergroup num_edn_reqs_cg with function sample(int num_edn_reqs);
     num_edn_reqs_cp: coverpoint num_edn_reqs {
-      bins less_then_five_reqs = {[1:4]};
+      bins less_than_five_reqs = {[1:4]};
       bins five_or_more_reqs   = {[5:$]};
     }
   endgroup
 
+  covergroup num_checked_pings_cg with function sample (int num_pings);
```
Comment (sriyerg) [inline hw/ip_templates/alert_handler/dv/env/alert_handler_env_cov.sv:289]: I dont see this added to 
https://docs.opentitan.org/hw/top_earlgrey/ip_autogen/alert_handler/doc/dv/index.html#testplan

A brief description on what it captures will be nice (or, add it to the testplan). I see you are covering the number of pings. But are you doing this within a specific window of time (I don't think so), or for the entire duration of the test? I wonder if it is more useful to cover the number of cycles between pings instead (but I could be wrong!).

**Review by sriyerg — APPROVED** [review]
Left a question - otherwise, LGTM!

File: hw/ip_templates/alert_handler/dv/env/alert_handler_env_cov.sv (hunk 2 above)
Comment (cindychip) [reply hw/ip_templates/alert_handler/dv/env/alert_handler_env_cov.sv:289]: Oh thanks for catching that. I forget to add it to the testplan.
As for the coverbins - I mainly want to ensure the simulation is long enough to capture more than one ping timing checks. 
As for the number of cycles between pings - I think that is a good point and I can add another coverpoint for it!
Thanks for the review!

File: hw/ip_templates/alert_handler/dv/env/alert_handler_env_cov.sv [hunk 3]
```diff
@@ -281,11 +281,31 @@ class alert_handler_env_cov extends cip_base_env_cov #(.CFG_T(alert_handler_env_
...
+      bins less_than_ten_pings = {[1:9]};
+      bins ten_to_twenty_pings = {[10:19]};
+      bins more_than_twenty_pings = {[20:$]};
+    }
+  endgroup
+
+  covergroup cycles_between_pings_cg with function sample (int num_cycles_between_pings);
+    num_cycles_cp: coverpoint num_cycles_between_pings{
+      bins less_than_5000_cycs = {[1:4_999]};
+      bins less_than_10000_cycs = {[5_000:99_999]};
```
Comment (sriyerg) [inline hw/ip_templates/alert_handler/dv/env/alert_handler_env_cov.sv:300]: 100k v/s 10k? Same below...

File: hw/ip_templates/alert_handler/dv/env/alert_handler_env_cov.sv (hunk 2 above)
Comment (sriyerg) [reply hw/ip_templates/alert_handler/dv/env/alert_handler_env_cov.sv:289]: Ohh wait - add it only if it really makes sense!

File: hw/ip_templates/alert_handler/dv/env/alert_handler_env_pkg.sv [hunk 4]
```diff
@@ -42,6 +42,8 @@ package alert_handler_env_pkg;
   parameter uint NUM_CRASHDUMP             = NUM_ALERT_CLASSES * (alert_handler_reg_pkg::AccuCntDw
                                              + alert_handler_reg_pkg::EscCntDw + 3) +
                                              NUM_ALERTS + NUM_LOCAL_ALERTS;
+  parameter bit[15:0] MAX_PING_WAIT_CYCLES = '1;
```
Comment (sriyerg) [inline hw/ip_templates/alert_handler/dv/env/alert_handler_env_pkg.sv:45]: Seems unused?

File: hw/ip_templates/alert_handler/dv/env/alert_handler_scoreboard.sv [hunk 5]
```diff
@@ -563,6 +571,55 @@ class alert_handler_scoreboard extends cip_base_scoreboard #(
...
+            // randomly selected and we can not guarantee the random alert index is valid (exists),
+            // enabld, and locked. However, ping timer should always be sent out. So the max wait
+            // time in doubled.
+            int ping_wait_time_ns = (cfg.clk_rst_vif.clk_period_ps / 1_000 + 1) * 16'hFFFF * 2;
+
+            int ping_wait_cycs;
+            fork begin : isolation_fork
+              fork
+                begin
+                  while (ping_wait_cycs <= MAX_PING_TIMEOUT_CYCLES) begin
```
Comment (sriyerg) [inline hw/ip_templates/alert_handler/dv/env/alert_handler_scoreboard.sv:597]: You mean `MAX_PING_WAIT_CYCLES`...

File: hw/ip_templates/alert_handler/dv/env/alert_handler_scoreboard.sv [hunk 6]
```diff
@@ -563,6 +571,55 @@ class alert_handler_scoreboard extends cip_base_scoreboard #(
...
+                    ping_wait_cycs++;
+                  end
+                  `uvm_fatal(`gfn, "expect to see a ping request!");
+                end
+                begin
+                  wait(ping_triggered);
+                end
+              join_any
+              disable fork;
+            end join
```
Comment (sriyerg) [inline hw/ip_templates/alert_handler/dv/env/alert_handler_scoreboard.sv:608]: Could you refactor this into a separate task?

File: hw/ip_templates/alert_handler/dv/env/alert_handler_scoreboard.sv [hunk 7]
```diff
@@ -563,6 +571,55 @@ class alert_handler_scoreboard extends cip_base_scoreboard #(
...
+
+            int ping_wait_cycs;
+            fork begin : isolation_fork
+              fork
+                begin
+                  while (ping_wait_cycs <= MAX_PING_TIMEOUT_CYCLES) begin
+                    cfg.clk_rst_vif.wait_clks(1);
+                    ping_wait_cycs++;
+                  end
+                  `uvm_fatal(`gfn, "expect to see a ping request!");
```
Comment (sriyerg) [inline hw/ip_templates/alert_handler/dv/env/alert_handler_scoreboard.sv:601]: Hmmm...this will never fire! The `while` loop above has an exit condition that will never be met.

Comment (sriyerg) [inline hw/ip_templates/alert_handler/dv/env/alert_handler_scoreboard.sv:601]: Perhaps reword this to `Timeout occurred waiting for a ping`? Is it really a fatal error?

File: hw/ip_templates/alert_handler/dv/env/alert_handler_scoreboard.sv [hunk 8]
```diff
@@ -563,6 +571,55 @@ class alert_handler_scoreboard extends cip_base_scoreboard #(
...
+            ping_timer_en = 0;
+            num_checked_pings = 0;
+          end
+          begin
+            // The ping timer is 16 bits so ideally we should see alert_ping -> esc_ping ->
+            // alert_ping ... with the max length of 16'hFFFF clock cycle. However alert_ping is
+            // randomly selected and we can not guarantee the random alert index is valid (exists),
+            // enabld, and locked. However, ping timer should always be sent out. So the max wait
+            // time in doubled.
+            int ping_wait_time_ns = (cfg.clk_rst_vif.clk_period_ps / 1_000 + 1) * 16'hFFFF * 2;
```
Comment (sriyerg) [inline hw/ip_templates/alert_handler/dv/env/alert_handler_scoreboard.sv:591]: Now this seems unused.

File: hw/ip_templates/alert_handler/dv/env/alert_handler_scoreboard.sv (hunk 5 above)
Comment (cindychip) [reply hw/ip_templates/alert_handler/dv/env/alert_handler_scoreboard.sv]: sorry about that, somehow my force push did not take the latest version..

File: hw/ip_templates/alert_handler/dv/env/alert_handler_scoreboard.sv (hunk 8 above)
Comment (cindychip) [reply hw/ip_templates/alert_handler/dv/env/alert_handler_scoreboard.sv]: sorry about that, somehow my force push did not take the latest version..

File: hw/ip_templates/alert_handler/dv/env/alert_handler_env_cov.sv (hunk 3 above)
Comment (cindychip) [reply hw/ip_templates/alert_handler/dv/env/alert_handler_env_cov.sv:300]: Good catch, thanks I will update the namings.

File: hw/ip_templates/alert_handler/dv/env/alert_handler_env_cov.sv (hunk 2 above)
Comment (cindychip) [reply hw/ip_templates/alert_handler/dv/env/alert_handler_env_cov.sv:289]: Thanks Sri :) I think it is nice of have - in that way we can have more evidence that LFSR outputs random values.

File: hw/ip_templates/alert_handler/dv/env/alert_handler_env_pkg.sv (hunk 4 above)
Comment (cindychip) [reply hw/ip_templates/alert_handler/dv/env/alert_handler_env_pkg.sv:45]: Good catch, I will clean this up.

File: hw/ip_templates/alert_handler/dv/env/alert_handler_scoreboard.sv (hunk 6 above)
Comment (cindychip) [reply hw/ip_templates/alert_handler/dv/env/alert_handler_scoreboard.sv]: oh that is a good point, will do!

File: hw/ip_templates/alert_handler/dv/env/alert_handler_env_pkg.sv (hunk 4 above)
Comment (cindychip) [reply hw/ip_templates/alert_handler/dv/env/alert_handler_env_pkg.sv:44]: Yeah because the previous code checked-in a wrong version. I updated the code below to use the correct parameter name.
